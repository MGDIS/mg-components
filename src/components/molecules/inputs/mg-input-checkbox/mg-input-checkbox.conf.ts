/**
 * type CheckboxItem
 * use to match checkbox attributes
 */
export type CheckboxItem = {
  id: string;
  title: string;
  value: boolean | null;
  disabled?: boolean;
};

/**
 * type CheckboxValue
 * use to match returned value
 */
export type CheckboxValue = {
  title: string;
  value: boolean | null;
  disabled?: boolean;
};
