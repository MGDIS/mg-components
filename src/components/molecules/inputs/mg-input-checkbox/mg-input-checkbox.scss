@import '../mg-input.scss';
@import '../../../../styles/a11y.scss';
