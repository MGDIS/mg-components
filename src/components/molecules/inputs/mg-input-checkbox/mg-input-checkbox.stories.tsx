import { h } from '@stencil/core';

export default {
  component: 'mg-input-checkbox',
  title: 'Molecules/Inputs/mg-input-checkbox',
  parameters: { actions: { handles: ['value-change'] } },
};

/**
 * 1. camelCase arguments must be written in the template, for exemple labelOnTop must be placed in the template as label-on-top={args.labelOnTop}
 * 2. boolean arguments with a default true value must be added like display-character-left={args.displayCharacterLeft ? 'true' : 'false'}
 *
 * @param {any} args component arguments
 * @returns {HTMLElement} HTMLElement
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const Template = (args: any): HTMLElement => {
  // Extract slot so it won't be render as an attribute
  const labelOnTop = args.labelOnTop;
  delete args.labelOnTop;
  const helpText = args.helpText;
  delete args.helpText;
  const inputVerticalList = args.inputVerticalList;
  delete args.inputVerticalList;
  const labelHide = args.labelHide;
  delete args.labelHide;
  // return element
  return <mg-input-checkbox {...args} label-on-top={labelOnTop} label-hide={labelHide} help-text={helpText} input-vertical-list={inputVerticalList}></mg-input-checkbox>;
};

export const MgInputCheckbox = Template.bind({});
MgInputCheckbox.args = {
  // Global
  value: [
    {
      title: 'HT',
      value: true,
    },
    {
      title: 'TTC',
      value: false,
      disabled: true,
    },
    {
      title: 'Mixte (HT/TTC)',
      value: null,
    },
  ],
  identifier: 'identifier',
  name: 'input-name',
  // Label
  label: 'Option',
  labelOnTop: false,
  labelHide: false,
  // placement
  inputVerticalList: false,
  // Input
  required: false,
  disabled: false,
  readonly: false,
  // Tooltip
  tooltip: 'This is a tooltip',
  // Help Text
  helpText: 'Help text with html <strong>bold</strong>, <em>italic</em>.',
};
