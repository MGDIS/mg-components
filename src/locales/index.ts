import en from './en/messages.json';
import fr from './fr/messages.json';

const defaultLocale = 'en';
const messages = { en, fr };

/**
 * Get locale
 *
 * @param {HTMLElement} element element that needs to know the locale to use
 * @returns {string} locale
 */
const getLocale = (element: HTMLElement): string => {
  return (element.closest('[lang]') as HTMLElement)?.lang || navigator.language || defaultLocale;
};

/**
 * Get Intl object
 * We load the defined locale but for now we only support the first subtag for messages
 *
 * @param {HTMLElement} element element we need to get the language
 * @returns {{ locale: string; messages: unknown }} messages object
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const initLocales = (element: HTMLElement): { locale: string; messages: any } => {
  // Get local
  const locale = getLocale(element);
  // Only keep first subtag
  const localeSubtag = locale.split('-').shift();
  // Return
  return {
    locale,
    messages: messages[localeSubtag] || messages[defaultLocale],
  };
};
