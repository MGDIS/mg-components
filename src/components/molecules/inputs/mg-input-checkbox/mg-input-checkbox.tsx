import { Component, Event, h, Prop, EventEmitter, State, Watch, Element } from '@stencil/core';
import { MgInput } from '../MgInput';
import { createID, ClassList } from '../../../../utils/components.utils';
import { messages } from '../../../../locales';
import { CheckboxOption, CheckboxValue }from '../../../../types/components.types';

/**
* type CheckboxOption validation function
* @param CheckboxOption
* @returns {boolean}
*/
const isCheckboxOption = (option: CheckboxOption): boolean => typeof option === 'object' && typeof option.title === 'string' && (option.value === null || typeof option.value === 'boolean') && option.value !== undefined;

@Component({
  tag: 'mg-input-checkbox',
  styleUrl: 'mg-input-checkbox.scss',
  shadow: true,
})

export class MgInputCheckbox {

  /************
   * Internal *
   ************/
   private classError = 'is-not-valid';

  /**************
  * Decorators *
  **************/

  /**
  * Component value
  * If item.value is `null`, checkbox will be indeterminate by default
  * Required
  */
  @Prop({ mutable:true, reflect: true }) value!: CheckboxValue[];
  @Watch('value')
  validateValue(newValue){
    if(newValue && (newValue as Array<CheckboxOption>).every(item => isCheckboxOption(item))) {
      this.options = newValue.map((item, index) => ({
        id: `${this.identifier}_${index.toString()}`,
        title:item.title,
        value:item.value,
        disabled: item.disabled
      }));
    }
    else {
      throw new Error('<mg-input-checkbox> prop "value" is required and all values must be the same type, CheckboxOption.')
    }
  }

  /**
  * Identifier is used for the element ID (id is a reserved prop in Stencil.js)
  * If not set, it will be created.
  */
  @Prop() identifier?: string = createID('mg-input-checkbox');

  /**
  * Input name
  * If not set the value equals the identifier
  */
  @Prop() name?: string = this.identifier;

  /**
  * Input label
  * Required
  */
  @Prop() label!: string;

  /**
  * Define if label is displayed on top
  */
  @Prop() labelOnTop: boolean;

  /**
  * Define if label is visible
  */
  @Prop() labelHide: boolean = false;

  /**
  * Define if inputs are display verticaly
  */
  @Prop() inputVerticalList: boolean = false;

  /**
  * Define if input is required
  */
  @Prop() required: boolean = false;

  /**
  * Define if input is readonly
  */
  @Prop() readonly: boolean = false;

  /**
  * Define if input is disabled
  */
  @Prop() disabled: boolean = false;

  /**
  * Add a tooltip message next to the input
  */
  @Prop() tooltip: string;

  /**
  * Template to use for characters left sentence
  */
  @Prop() helpText: string;

  /**
  * Force valid component
  */
  @Prop({ mutable: true, reflect: true }) valid: boolean;

  /**
  * Force invalid component
  */
  @Prop({ mutable: true, reflect: true }) invalid: boolean;

  /**
  * Component classes
  */
  @State() classList: ClassList = new ClassList(['mg-input--checkbox']);

  /**
  * Error message to display
  */
  @State() errorMessage: string;

  /**
  * Formated value for display
  */
  @State() options: (CheckboxOption)[];

  /**
  * Emitted event when value change
  */
  @Event() valueChange: EventEmitter<CheckboxValue[]>

  /**
  * Get component DOM element
  */
  @Element() element: HTMLMgInputCheckboxElement;

  /**
  * Handle input event
  * @param event
  */
  private handleInput = (event:InputEvent & { target: HTMLInputElement }) => {
    this.options = this.options.map(option => {
      if (option.id === event.target.id) {
        option.value = Boolean(event.target.checked);
      }
      return option;
    })

    this.value = this.options.map(o => ({value: o.value, title: o.title}));
    this.valueChange.emit(this.value);
  }

  /**
   * Handle blur event
   * @param event
   */
  private handleBlur = (event:FocusEvent) => {
    // Check validity
    this.checkValidity(event.target);
  }

  /**
  * Check if input is valid
  * @param element
  */
  private checkValidity(element) {
    // we check element  OR group validity
    const elementValidity = element.checkValidity();
    const validity = elementValidity || this.checkInputGroupValidity();

    // Set error message
    this.errorMessage = undefined;
    if(!validity && element.validity.valueMissing) {
      this.errorMessage = messages.errors.required;
    }

    // Set validity
    this.valid = validity;
    this.invalid = !validity;

    // Update class
    if(validity) {
      this.classList.delete(this.classError);
    }
    else {
      this.classList.add(this.classError);
    }
  }

  /**
   * check input group validity
   */
  private checkInputGroupValidity(): boolean {
    const inputs = Array.from(this.element.shadowRoot.querySelectorAll('input[type=checkbox]'));
    const getValidInput = inputs.find(i => {
      const element = i as HTMLInputElement;
      // we skip disabled elements
      if (element.disabled) return false;

      return element.checkValidity();
    });
    return getValidInput !== undefined;
  }

  /*************
  * Lifecycle *
  *************/

  componentWillLoad() {
    // Check values format
    this.validateValue(this.value);
  }

  render() {
    return (
      <MgInput
        identifier={this.identifier}
        classList={this.classList}
        label={this.label}
        labelOnTop={this.labelOnTop}
        labelHide={this.labelHide}
        required={this.required}
        readonly={false}
        value={this.value && this.value.toString()}
        readonlyValue={undefined}
        tooltip={this.tooltip}
        displayCharacterLeft={undefined}
        characterLeftTemplate={undefined}
        maxlength={undefined}
        helpText={this.helpText}
        errorMessage={this.errorMessage}
        isFieldset={true}
      >
        <ul class={"mg-input__input-group-container mg-input__input-group-container--checkbox " + (this.inputVerticalList ? 'mg-input__input-group-container--vertical' : '')}>
          {this.options
            .filter((item) => {
              return !this.readonly || item.value;
            }).map(input => (
            <li class="mg-input__input-group">
              <input
                type="checkbox"
                id={input.id}
                name={this.identifier}
                value={input.value && input.value.toString()}
                checked={Boolean(input.value)}
                disabled={this.readonly || this.disabled || input.disabled}
                required={this.required}
                indeterminate={input.value === null}
                onInput={this.handleInput}
                onBlur={this.handleBlur}
              />
              <label htmlFor={input.id}>{input.title}</label>
            </li>
          ))}
        </ul>
      </MgInput>
    );
  }

}
