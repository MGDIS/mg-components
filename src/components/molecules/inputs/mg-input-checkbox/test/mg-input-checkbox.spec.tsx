import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgInputCheckbox } from '../mg-input-checkbox';
import { messages } from '../../../../../locales';

const getPage = (args) => newSpecPage({
  components: [MgInputCheckbox],
  template: () => (<mg-input-checkbox {...args}></mg-input-checkbox>)
});

describe('mg-input-checkbox', () => {
  test.each([
    {label: 'label', identifier: "identifier"},
    {label: 'label', identifier: "identifier", labelOnTop: true},
    {label: 'label', identifier: "identifier", readonly: true},
    {label: 'label', identifier: "identifier", readonly: true, value: true},
    {label: 'label', identifier: "identifier", readonly: true, value: false},
    {label: 'label', identifier: "identifier", tooltip: "My Tooltip Message"},
    {label: 'label', identifier: "identifier", tooltip: "My Tooltip Message", indeterminate: true},
    {label: 'label', identifier: "identifier", tooltip: "My Tooltip Message", indeterminate: true, readonly: true},
    {label: 'label', identifier: "identifier", tooltip: "My Tooltip Message", indeterminate: true, value: null},
  ])('Should render with args %s:', async (args) => {
    const { root } = await getPage(args);
    expect(root).toMatchSnapshot();
  });

  test.each(["", undefined])('Should not render with invalid label property : %s', async (value) => {
    try {
      await getPage({label:value});
    }
    catch (err) {
      expect(err.message).toMatch('prop "label" is required')
    }
  });

  test('Should trigger events', async ()=> {
    const inputValue = false;
    const args = {label: 'label', identifier: "identifier", helpText: "My help text"};
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-checkbox');
    const input = element.shadowRoot.querySelector('input');

    //mock validity
    input.checkValidity = jest.fn(()=> true);
    Object.defineProperty(input, 'validity', { get: jest.fn(()=> ({
      valueMissing: false
    }))});

    jest.spyOn(page.rootInstance.valueChange, 'emit');

    input.dispatchEvent(new CustomEvent('focus', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.classList.has('is-focused')).toBeTruthy();

    expect(page.root).toMatchSnapshot(); //Snapshot on focus

    input.value = inputValue.toString();
    input.dispatchEvent(new CustomEvent('input', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.valueChange.emit).toHaveBeenCalledWith(inputValue);

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();
    expect(page.rootInstance.classList.has('is-focused')).toBeFalsy();
  });

  test.each([
    {validity: true, valueMissing: false},
    {validity: false, valueMissing: true},
    {validity: false, valueMissing: false, value:"Blu"},
  ])('validity (%s), valueMissing (%s)', async ({validity, valueMissing, value})=> {
    const args = {label: 'label', identifier: "identifier", value};
    const page = await getPage(args);

    const element = page.doc.querySelector('mg-input-checkbox');
    const input = element.shadowRoot.querySelector('input');

    //mock validity
    input.checkValidity = jest.fn(()=> validity);
    Object.defineProperty(input, 'validity', { get: jest.fn(()=> ({
      valueMissing,
    }))});

    input.dispatchEvent(new CustomEvent('blur', { bubbles: true }));
    await page.waitForChanges();

    if(validity) {
      expect(page.rootInstance.errorMessage).toBeUndefined();
    }
    else if (valueMissing){
      expect(page.rootInstance.errorMessage).toEqual(messages.errors.required);
    }
    expect(page.rootInstance.valid).toEqual(validity);
    expect(page.rootInstance.invalid).toEqual(!validity);
  });

});

