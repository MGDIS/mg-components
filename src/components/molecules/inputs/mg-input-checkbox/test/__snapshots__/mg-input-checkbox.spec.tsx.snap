// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  disabled: true
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" disabled="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  helpText: 'Hello jocker'
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
        <div class="mg-input__input-container__help-text" id="identifier-help-text">
          Hello jocker
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  inputVerticalList: true
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox mg-input__input-group-container--vertical">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  labelHide: true
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title class="sr-only" identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  labelOnTop: true
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox mg-input--label-on-top">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  readonly: true
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" disabled="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  readonly: true
}: 2`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" disabled="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  required: true
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="" required="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" id="identifier_0" name="identifier" required="" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" required="" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_2" name="identifier" required="" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_3" indeterminate="" name="identifier" required="" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args {
  label: 'label',
  identifier: 'identifier',
  value: [Array],
  tooltip: 'Batman is a DC Comics license'
}: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
          <mg-tooltip identifier="identifier-tooltip" message="Batman is a DC Comics license">
            <mg-icon icon="info"></mg-icon>
          </mg-tooltip>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should render with args { label: 'label', identifier: 'identifier', value: [Array] }: 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should trigger events, case validity check false 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
        <div class="mg-input__input-container__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should trigger events, case validity check false 2`] = `
<mg-input-checkbox invalid="">
  <mock:shadow-root>
    <fieldset class="is-not-valid mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text identifier-error" id="identifier_0" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text identifier-error" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text identifier-error" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text identifier-error" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
        <div class="mg-input__input-container__help-text" id="identifier-help-text">
          My help text
        </div>
        <div aria-live="assertive" class="mg-input__input-container__error" id="identifier-error">
          Ce champs est obligatoire.
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should trigger events, case validity check true 1`] = `
<mg-input-checkbox>
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" disabled="" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_2" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
        <div class="mg-input__input-container__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;

exports[`mg-input-checkbox Should trigger events, case validity check true 2`] = `
<mg-input-checkbox valid="">
  <mock:shadow-root>
    <fieldset class="mg-input mg-input--checkbox">
      <mg-input-title identifier="identifier" is-legend="">
        label
      </mg-input-title>
      <div class="mg-input__input-container">
        <div class="mg-input__input-container__input">
          <ul class="mg-input__input-group-container mg-input__input-group-container--checkbox">
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" checked="" id="identifier_0" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_0">
                batman
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_1" name="identifier" type="checkbox" value="false">
              <label htmlfor="identifier_1">
                robin
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" checked="" id="identifier_2" name="identifier" type="checkbox" value="true">
              <label htmlfor="identifier_2">
                jocker
              </label>
            </li>
            <li class="mg-input__input-group">
              <input aria-describedby="identifier-help-text" id="identifier_3" indeterminate="" name="identifier" type="checkbox">
              <label htmlfor="identifier_3">
                bane
              </label>
            </li>
          </ul>
        </div>
        <div class="mg-input__input-container__help-text" id="identifier-help-text">
          My help text
        </div>
      </div>
    </fieldset>
  </mock:shadow-root>
</mg-input-checkbox>
`;
